### DISCUSSION TITLE
Event handler dependencies

### DISCUSSION THREAD

**Discussion by slashdotdash** [commanded/commanded#573] [Ideas]
Allow an event handler to have a dependency upon one or more other handlers. This would ensure that any prerequisite event handlers have successfully processed an event before the dependent handler.

### Proposal

```elixir
defmodule PrerequisiteHandler do
  use Commanded.Event.Handler, name: __MODULE__

  def handle(%AnEvent{..}, metadata) do
    # This `handle/2` function always executes before the dependent handler for the given event
    :ok
  end
end

defmodule DependentHandler do
  use Commanded.Event.Handler, name: __MODULE__, dependencies: [PrerequisiteHandler]

  def handle(%AnEvent{..}, metadata) do
    # This `handle/2` function always executes after the prerequisite handler for the given event
    :ok
  end
end
```

The `dependencies` config option would allow a list of one or more prerequisite event handlers to be configured. 

It should *not* be possible to configure circular dependencies (e.g. handler A depends on handler B which depends on handler A).

Consideration must be made when a prerequisite handler fails to handle an event and stops. What should happen to the dependent handler (e.g. timeout, error callback).

**Comment by slashdotdash** [comment]
For now it's possible to implement your own function that takes advantage of Commanded's subscriptions to have one handler wait for another handler.

```elixir
defmodule Handler do
  def wait_for(metadata, handler, timeout \\ 5_000) do
    stream_id = Map.get(metadata, :stream_id)
    stream_version = Map.get(metadata, :stream_version)

    Commanded.Subscriptions.wait_for(
      stream_id,
      stream_version,
      [consistency: [handler]],
      timeout
    )
  end
end
```

### Usage

```elixir
def handle(%Event{}, metadata) do
  :ok = Handler.wait_for(metadata, SomeOtherHandler)

  # `SomeOtherHandler` has now successfully processed this event

  :ok
end
```

**Comment by trbngr** [comment]
will this work if multiple handlers depend on the same dependent handler?

**Comment by Freyskeyd** [comment]
Here comes the circular dependencies :) 

I don't really see any use case (in my personnal scope) to build a dependency tree for event handlers. If an event handler need to wait for another one to finish it's mostly a design issue in my mind. The waiting handler should be part of the first one.

Maybe I'm missing some part of the proposal.

**Comment by justgage** [comment]
This would be very useful to me, mainly because I have a projector per table/store. And because there's a foreign key between them it's kind of important for one to wait for the other. Yes this might be a bad design, but it seems pretty reasonable to me. This allows one event handler to run as fast as it can and the other to lag behind it.

**Comment by yordis** [comment]
> Yes this might be a bad design, but it seems pretty reasonable to me. 

I hope it is no longer the case.

> And because there's a foreign key between them it's kind of important for one to wait for the other. 

That use case that you shared is the perfect example of what not to do (have dependencies between your read-models)

---- 

This feature, more often than not, comes from bad system design and taking the suboptimal decisions from my experience, and it is not hard to understand why people would like to have something like that; but still, it only causes problems and messy software in the long term.

I wouldn't like to see something like this be encouraged.
